core/fault/stacktrace/pack.go

```diff
@@ -0,0 +1,374 @@
+	for ; c > 0; c-- {
+		fmt.Printf("%v", v&1)
+		v >>= 1
+	}
+}
+func (bitPrinter) WriteBit(v uint64) {
+	if v == 0 {
+		fmt.Print("⁰")
+	} else {
+		fmt.Print("¹")
+	}
+}
+
+type bitwriter interface {
+	Write(bits uint64, count uint32)
+	WriteBit(bit uint64)
+}
+
+func writeVLE(i uint64, w bitwriter) {
+	chunkSize := uint32(bitChunkMinSize)
+	for i != 0 {
+		w.WriteBit(0)
+		w.Write(uint64(i), chunkSize)
+		i >>= chunkSize
+		chunkSize += bitChunkIncrement
+	}
+	w.WriteBit(1)
+}
+
+func readVLE(bs *binary.BitStream) (uint64, bool) {
+	chunkSize := uint32(bitChunkMinSize)
+	i, s := uint64(0), uint32(0)
+	for {
+		if !bs.CanRead(1) {
+			return 0, false
+		}
+		if bs.ReadBit() == 1 {
+			break
+		}
+		if !bs.CanRead(chunkSize) {
+			return 0, false
+		}
+		i |= bs.Read(chunkSize) << s
+		s += chunkSize
+		chunkSize += bitChunkIncrement
+	}
+	return uint64(i), true
+}
+
+type compressor interface {
+	compress([]uint64) []uint64
+	decompress([]uint64) []uint64
+}
+
+type noCompressor struct{}
+
+func (p noCompressor) compress(c []uint64) []uint64 {
+	return c
+}
+
+func (p noCompressor) decompress(c []uint64) []uint64 {
+	return c
+}
+
+type baseCompressor struct{}
+
+func (p baseCompressor) compress(c []uint64) []uint64 {
+	base := ^(uint64)(0)
+	for _, v := range c {
+		if v < base {
+			base = v
+		}
+	}
+	out := make([]uint64, len(c)+1)
+	out[0] = base
+	for i, v := range c {
+		out[i+1] = v - base
+	}
+	return out
+}
+
+func (p baseCompressor) decompress(c []uint64) []uint64 {
+	if len(c) < 2 {
+		return c
+	}
+	out := make([]uint64, len(c)-1)
+	base := c[0]
+	for i, pc := range c[1:] {
+		out[i] = pc + base
+	}
+	return out
+}
+
+type xorCompressor struct{}
+
+func (p xorCompressor) compress(c []uint64) []uint64 {
+	var prev uint64
+	out := make([]uint64, len(c))
+	for i, v := range c {
+		out[i], prev = v^prev, v
+	}
+	return out
+}
+
+func (p xorCompressor) decompress(c []uint64) []uint64 {
+	var prev uint64
+	out := make([]uint64, len(c))
+	for i, val := range c {
+		v := val ^ prev
+		out[i], prev = v, v
+	}
+	return out
+}
+
+type backRefCompressor struct {
+	pack    func(ref, val uint64) uint64
+	unpack  func(ref, packed uint64) uint64
+	history int
+}
+
+func packXor(ref, val uint64) uint64      { return ref ^ val }
+func unpackXor(ref, packed uint64) uint64 { return ref ^ packed }
+
+func packDiff(ref, val uint64) uint64 {
+	if ref > val {
+		return (ref - val) << 1
+	}
+	return (val-ref)<<1 | 1
+}
+func unpackDiff(ref, val uint64) uint64 {
+	sign, val := val&1, val>>1
+	if sign == 0 {
+		return ref + val
+	}
+	return ref - val
+}
+
+func (p backRefCompressor) compress(c []uint64) []uint64 {
+	out := make([]uint64, 0, len(c))
+
+	size := func(val uint64) uint32 {
+		bs := binary.BitStream{}
+		writeVLE(val, &bs)
+		return bs.WritePos
+	}
+
+	for i, v := range c {
+		bestValue := v
+		bestIndex := 0
+		bestScore := size(uint64(bestIndex)) + size(bestValue)
+		for idx := 1; idx < p.history; idx++ {
```

Comment:
I do not get this.  Why not just: for idx := 0; idx < i; idx++ { ...

Reply:
The index used to be fixed size - you right in that the history length is now dynamic, so yes, this is a good suggestion.

Reply:
Done